### PR TITLE
 HARP-6510 Running rendering tests locally in docker container. 

### PR DESCRIPTION
**Pull request by arekspiewak1991**
HARP-6510 This change allows to run rendering tests locally in docker container.

- add script `rendering-tests-reference-platform.sh` for running tests with specified reference platform (`Chrome Headless with software renderer / Linux`) 
- adds possibility to inject module as express middleware to existing express app instance in: `with-http-server.ts`
- uses image from https://github.com/SeleniumHQ/docker-selenium

### CODE REVIEW

**Comment by arekspiewak1991** [comment]
@heremaps-bot retest this please

**Comment by arekspiewak1991** [comment]
@heremaps-bot retest this please

**Comment by arekspiewak1991** [comment]
@heremaps-bot retest this please

File: @here/harp-test-utils/lib/rendering/ReferenceImageLocator.ts [hunk 1]
```diff
@@ -22,18 +22,18 @@ export function getReferenceImageUrl(imageProps: TestImageProps) {
...
 export function defaultReferenceImageResolver(imageProps: TestImageProps) {
     if (imageProps.name && imageProps.module) {
-        const baseUrl = "../ibct-results/";
-
         const platform = getOverride("IBCT_REFERENCE_PLATFORM", imageProps.platform || "");
-        const platformPart = platform === "" ? "" : `${platform}/`;
-        const modulePart = imageProps.module ? imageProps.module + "-" : "";
         const extra = imageProps.extra || "";
-        return `${baseUrl}${platformPart}${modulePart}${imageProps.name}${extra}.png`;
+
```
Comment (zbigg) [inline @here/harp-test-utils/lib/rendering/ReferenceImageLocator.ts]: const queryString = Object.keys(imageProps).map(key => `key=${urlencode(imageProps[key])}`).join("&")

File: test/README.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+### Prerequisites
+-   Linux
+-   Docker installed
+
+### Tests execution - build and run locally
+
+1. Run command:
+
+```shell
+    bash scripts/docker-tests-runner.sh
```
Comment (zbigg) [inline test/README.md]: It should be something like "yarn rendering-tests".
Scripts should be hidden.

File: test/README.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
+# Performance Tests
+
+# Rendering Tests
+
+Tests for rendering components, which compares image result and stored reference image.
```
Comment (zbigg) [inline test/README.md]: Reword.
Tests for components *that* render images. They render images to canvas and compare the result image with approved reference images.

File: test/README.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+```shell
+    bash scripts/docker-tests-runner.sh
+    #OR
+    ./scripts/docker-tests-runner.sh
+```
+ 
+2. Local reference images
+
+Now, the problem is that your browser platform string doesn't look like the one that CI uses
```
Comment (zbigg) [inline test/README.md]: This text doesn't apply as we don't have CI.
You should describe local flow only, where first run just skips all tests because there are no reference images.

File: @here/harp-test-utils/lib/rendering/RenderingTestResultServer.ts [hunk 5]
```diff
@@ -158,21 +165,56 @@ export function installMiddleware(app: express.Router, basePath: string) {
     });
 
     outputBasePath = basePath;
+    logger.log("installMiddleware output basepath:", outputBasePath);
+
     const jsonParser = bodyParser.json({ limit: 1024 * 1024 * 16 });
     app.get("/ibct-report", jsonParser, getIbctReport);
     app.post("/ibct-feedback", jsonParser, postIbctFeedback);
+    app.get("/reference-image", getReferenceImage);
```
Comment (zbigg) [inline @here/harp-test-utils/lib/rendering/RenderingTestResultServer.ts]: These changes should be in separate commits, but ok.
At least document all the featutures added in commit, because it's not only docker image.

File: scripts/docker-tests-runner.sh [hunk 6]
```diff
@@ -0,0 +1,11 @@
+#!/bin/sh
```
Comment (zbigg) [inline scripts/docker-tests-runner.sh]: Name is misleading.
It should be `rendering-tests-reference-platform`.

File: scripts/docker-tests-runner.sh [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+
+yarn build-tests
+
+npx ts-node -- \
```
Comment (zbigg) [inline scripts/docker-tests-runner.sh]: Need to create another script that will run these tests against tests served by `webpack-dev-server`.

Also `webpack.tests.config` are missing `RenderingTestResultServer` middleware.

File: test/README.md [hunk 8]
```diff
@@ -0,0 +1,86 @@
+# Rendering Tests
+
+Tests for components that render images. They render images to canvas and compare the result image
+with approved reference images.
+Tests are run in the browser headles mode, uses `mocha-webdriver-runner` to run `mocha` tests
```
Comment (musculman) [inline test/README.md]: Should be headless (not headles)

File: test/README.md [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+in browser context.
+
+## What to do first? Where to start?
+
+Before start the work on implementing new features you should run rendering tests first to
+generate actual images and store them as reference images to have possibility of making comparision
```
Comment (musculman) [inline test/README.md]: of making comparison

File: test/README.md [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+with right images after doing some changes in code.
+
+### Prerequisites
+-   Linux
```
Comment (musculman) [inline test/README.md]: Does it really work only on Linux? Is there a reason why the docker image does not work on windows/Mac?

File: package.json [hunk 11]
```diff
@@ -48,6 +49,9 @@
         "start-tests": "webpack-dev-server -d --config webpack.tests.config.js",
         "test-browser": "ts-node -- ./scripts/with-http-server.ts -C dist/test -p 8079 -- mocha-webdriver-runner http://localhost:8079/index.html",
         "build-tests": "webpack -d --config webpack.tests.config.js",
+        "rendering-tests": "./scripts/rendering-tests-reference-platform.sh",
```
Comment (musculman) [inline package.json]: Shouldn't this be run-rendering-tests?

File: test/README.md (hunk 10 above)
Comment (zbigg) [reply test/README.md]: There is no fundamental problem with running these on Windows/MacOS but it's not easy.

The man obstacle, is that we rely on `--network host` and it's [not available outside linux](https://docs.docker.com/network/host/).

Mac/Windows workflow would require lot of wiring, in order to open connection from  browser to test server running on localhost.

I think we need followup PR for this.

Some interesting links
 * https://stackoverflow.com/questions/40746453/how-to-connect-to-docker-host-from-container-on-windows-10-docker-for-windows
 * there is `host.docker.internal` -> https://docs.docker.com/docker-for-mac/networking & https://docs.docker.com/docker-for-windows/networking/

So seems, that with new dockers it's not so complex, but i would still propose another PR for this

Comment (musculman) [reply test/README.md]: Agreed with your proposal @zbigg , thanks for the explanation.

File: @here/harp-test-utils/lib/rendering/RenderingTestResultServer.ts (hunk 5 above)
Comment (arekspiewak1991) [reply @here/harp-test-utils/lib/rendering/RenderingTestResultServer.ts]: please verity (changed commit message and added one line of documentation... not sure if should I write something more there)

File: test/README.md (hunk 10 above)
Comment (arekspiewak1991) [reply test/README.md]: also - thanks Zbyszek answering that question...
I will try to add those changes with new PR.

File: scripts/docker-tests-runner.sh (hunk 7 above)
Comment (arekspiewak1991) [reply scripts/docker-tests-runner.sh]: I think done.

File: test/README.md (hunk 8 above)
Comment (arekspiewak1991) [reply test/README.md]: fixed

File: package.json (hunk 11 above)
Comment (arekspiewak1991) [reply package.json]: ok, I've changed that

File: package.json [hunk 12]
```diff
@@ -48,6 +49,9 @@
         "start-tests": "webpack-dev-server -d --config webpack.tests.config.js",
         "test-browser": "ts-node -- ./scripts/with-http-server.ts -C dist/test -p 8079 -- mocha-webdriver-runner http://localhost:8079/index.html",
         "build-tests": "webpack -d --config webpack.tests.config.js",
+        "run-rendering-tests": "./scripts/rendering-tests-reference-platform.sh",
+        "save-reference": "npx ts-node @here/harp-test-utils/lib/rendering/RenderingTestResultCli.ts save-reference",
```
Comment (musculman) [inline package.json]: I think the naming is not consistent as just by looking at the name you won't notice they are in fact used in combination.
What about?
* run-rendering-tests
* save-reference-rendering-tests
* approve-reference-rendering-tests

File: test/README.md [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+
+## Interactive mode
+
+1. Run command:
+
+```shell
+yarn start-tests
+```
+
+Starts `webpack-dev-server` which compiles your rendering tests on fly
```
Comment (musculman) [inline test/README.md]: on the fly

File: test/README.md (hunk 10 above)
Comment (musculman) [reply test/README.md]: I have tried testing this PR in linux wsl and mac and failed, on both systems with no clear error message. As expected I imagine, but I have not yet verified it with plain linux.

Comment (arekspiewak1991) [reply test/README.md]: @zbigg Could you please verify that again and test on your side?

Comment (zbigg) [reply test/README.md]: @musculman #1106 will adapt scripts and instructions, so all runs on mac + Docker fo Mac.

**Comment by arekspiewak1991** [comment]
@heremaps-bot retest this please

**Comment by arekspiewak1991** [comment]
@heremaps-bot retest this please
